app/controllers/users/sessions_controller.rb

```diff
@@ -56,10 +62,17 @@ def authenticate_with_otp_two_factor
 
     if user_params[:otp_attempt].present? && session[:otp_user_id]
       authenticate_user_with_otp_two_factor(user)
-    elsif user&.valid_password?(user_params[:password])
-      send_otp(user) if user.mobile_number
-
+    elsif user_params[:mobile_number].present? && session[:otp_user_id]
+      user.update!(mobile_number: user_params[:mobile_number], mobile_number_confirmed_at: DateTime.now)
+      send_otp(user)
       prompt_for_otp_two_factor(user)
+    elsif user&.valid_password?(user_params[:password])
+      if user.mobile_number.nil?
+        prompt_for_mobile_number(user)
+      else
+        send_otp(user)
+        prompt_for_otp_two_factor(user)
+      end
```

Comment:
This block is cyclomatic complexity hell. Can we make the conditionals read more naturally to indicate what we're doing?

Reply:
Suggestions are very welcome.